sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -81,6 +74,42 @@ trait StateStore {
     iterator()
   }
 
+  /**
+   * Return an iterator containing all the key-value pairs in the StateStore. Implementations must
+   * ensure that updates (puts, removes) can be made while iterating over this iterator.
+   */
+  def iterator(): Iterator[UnsafeRowPair]
+
+  /**
+   * Clean up the resource.
+   *
+   * The method name is to respect backward compatibility on [[StateStore]].
+   */
+  def abort(): Unit
+}
+
+/**
+ * Base trait for a versioned key-value store. Each instance of a `StateStore` represents a specific
+ * version of state data, and such instances are created through a [[StateStoreProvider]].
+ *
+ * Unlike [[ReadOnlyStateStore]], `abort` method may not be called if the `commit` method succeeds
+ * to commit the change. (`hasCommitted` returns `true`.) Otherwise, `abort` method will be called.
```

Comment:
It seems to me that `abort` has no clear semantics across `StateStore` and `ReadOnlyStateStore`.

Reply:
I agree, but not breaking backward compatibility is important, as I know several "ecosystem" implementations. That's why I added "different" javadoc between twos.

If I have a chance to redesign I would add close() as well, which will be called to clean up resources, regardless which methods was called (commit/abort) - so for ReadOnlyStateStore, there's no need to have abort() and only close() will be called, and for StateStore, after commit() and/or abort(), close() will be called too.

Reply:
If we really want to be sure, we can add `close()` as empty implementation by default, and let Spark guarantee to call the method regardless of calling commit() / abort(). For existing implementation, we can make close() to call abort() in WrappedReadStateStore, so that abort() is called once and clean up resource. Would it make sense to you?